### PR TITLE
json resource: Add handling for `command:` error

### PR DESCRIPTION
**Pull request by jerryaldrichiii**
This provides the user with messages if `json(command: 'command_that_errors')` exits non-zero and/or has STDERR.

Without this the resource will report "No output from command" when really the output is contained in STDERR.

### CODE REVIEW

File: lib/resources/json.rb [hunk 1]
```diff
@@ -96,10 +96,25 @@ def load_raw_from_file(path)
     end
 
     def load_raw_from_command(command)
-      command_output = inspec.command(command).stdout
-      raise Inspec::Exceptions::ResourceSkipped, "No output from command: #{command}" if command_output.nil? || command_output.empty?
+      command_result = inspec.command(command)
 
-      command_output
+      case command_result.exit_status
```
Comment (clintoncwolfe) [inline lib/resources/json.rb]: I don't think this should look at exit code at all. There are tools that indicate status by exit code, and non-zero does not mean "failure" (`grep` is an example).

File: lib/resources/json.rb [hunk 2]
```diff
@@ -96,10 +96,25 @@ def load_raw_from_file(path)
...
+      when 0
+
+        # TODO: Handle case where command exits 0 but STDERR is found.
+
+        if command_result.stdout.empty?
+          raise Inspec::Exceptions::ResourceSkipped, "No output from command: #{command}"
```
Comment (clintoncwolfe) [inline lib/resources/json.rb]: Might be good to leave a TODO here in each place you raise an exception - something like `TODO: Handle resource unable to run, see #3174"

File: test/unit/resources/json_test.rb [hunk 3]
```diff
@@ -100,18 +100,27 @@
       inspec.expects(:command).with(cmd_str).returns(command)
     end
 
-    it 'raises an exception if command stdout is nil' do
-      command.expects(:stdout).returns(nil)
+    it 'skips the resource if command exits zero and stdout is empty' do
```
Comment (clintoncwolfe) [inline test/unit/resources/json_test.rb]: As above, I think this is over-specified - just check for STDERR, and report the error code in the message.

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
Great catch! I think we shouldn't introduce checking for exit code, and just rely on the output.  Also small request to add some comments.

**Comment by jerryaldrichiii** [comment]
Hmmm...good point. I'll just include the exit code in any output/warnings.

EDIT: Thinking about this and there may be cases where STDERR is not empty but command exited 0. Examples like warn/debug messages (I think our json reporter does that even)

**Comment by clintoncwolfe** [comment]
The issue here is only around what to do when STDOUT is empty, not STDERR.

```
 if STDOUT is not empty
   parse it as JSON
 else
   Build an error message, using anything we can scrounge from stderr; include exit code as well
 end
```

**Comment by jerryaldrichiii** [comment]
Hey all, given a job change I won't be able to commit to finishing this PR in a timely manner.

Feel free to adopt :smile:

File: test/unit/resources/json_test.rb [hunk 4]
```diff
@@ -86,30 +86,61 @@
...
+      resource = run_json_cmd "echo 'not-valid-json'"
+
+      assert_resource_failed resource, "unexpected token at 'not-valid-json"
     end
 
-    it "returns the command output" do
-      command.expects(:stdout).returns("json goes here")
-      resource.send(:load_raw_from_command, cmd_str).must_equal "json goes here"
+    it "empty stdout, message in stderr" do
+      resource = run_json_cmd "echo bad args 1>&2 && false"
```
Comment (clintoncwolfe) [inline test/unit/resources/json_test.rb]: Looks like the `&&` breaks the windows tests...

**Review by clintoncwolfe — CHANGES_REQUESTED** [review]
Code looks great, just have a broken test to fix and then should be all set.

File: test/unit/resources/json_test.rb (hunk 4 above)
Comment (zenspider) [reply test/unit/resources/json_test.rb]: uuuugh. THAT thing again. I'll figure it out.

Comment (zenspider) [reply test/unit/resources/json_test.rb]: This is blocked by #4465, which I just filed.

**Review by miah — APPROVED** [review]

**Review by clintoncwolfe — APPROVED** [review]
Thanks!

**Pull request merged**